core/src/main/java/org/mapfish/print/output/ResourceBundleClassLoader.java

```diff
@@ -31,13 +33,24 @@ protected URL findResource(final String resource) {
 
   @Override
   public InputStream getResourceAsStream(final String resource) {
-    try {
-      final InputStream is = super.getResourceAsStream(resource);
-      byte[] bytes = new byte[is.available()];
-      is.read(bytes);
-      return new ByteArrayInputStream(new String(bytes, "utf-8").getBytes("iso-8859-1"));
+    ByteArrayOutputStream buffer;
+    try (InputStream is = super.getResourceAsStream(resource)) {
+      if (is == null) {
+        throw new IllegalArgumentException("Resource not found: " + resource);
+      }
+      buffer = new ByteArrayOutputStream();
+      int nRead;
+      byte[] data = new byte[1024];
+
+      while ((nRead = is.read(data, 0, data.length)) != -1) {
+        buffer.write(data, 0, nRead);
+      }
+
+      buffer.flush();
```

Comment:
I am not sure I understand the question. The buffer is used by line 54.

Reply:
Yes, but in the new code introduce a ByteArrayOutputStream and a while?
The old one was more direct...

Reply:
In the old code:

1. No null check: The super.getResourceAsStream(resource) may return a null if the resource does not exist. There's no null check, thus, it may lead to a NullPointerException at is.available().
2. Misuse of InputStream.available(): It's risky to rely on InputStream.available() for getting the size of the byte array. available() only returns an estimate, it's not guaranteed to be able to read that many bytes in all situations.
3. Incomplete read: is.read(bytes) may not read all the bytes. Its return value, indicating the number of bytes actually read, is being ignored.

So I did these changes but I left one last issue to keep the behavior unchanged:
Inefficient conversion of bytes to string and back to bytes: The conversion from bytes to String and back to bytes using different character sets can cause data loss, as all characters that can not be represented in ISO_8859_1 will be replaced.

Do you think I should fix that too ?

Reply:
> The conversion from bytes to String and back to bytes using different character sets can cause data loss, as all characters that can not be represented in ISO_8859_1 will be replaced.

We are reading localization files, then I think that no real need, and do we have a better possibility?

Reply:
I suggest we do no charset conversion. I will push a commit like that and we will see if we observe any side effects.

Reply:
No, the charset conversion is important because today we create only UTF-8 files,
And Java want's resource bundle in iso-8859-1, if it's no more the case we can drop the conversions :-)

Reply:
Note that there is a test for that:
https://github.com/mapfish/mapfish-print/blob/master/examples/src/test/resources/examples/resource_bundle/expected_output/requestData.png
:-)

Reply:
Since JavaSE 9 Properties files a loaded in UTF-8 : see [here](https://docs.oracle.com/javase/9/intl/internationalization-enhancements-jdk-9.htm#JSINT-GUID-5ED91AA9-B2E3-4E05-8E99-6A009D2B36AF)